services/web/server/src/simcore_service_webserver/meta_iterations.py

```diff
@@ -0,0 +1,332 @@
+        parameters_per_iter.append(parameters)
+        updated_nodes_per_iter.append(updated_nodes)
+
+    return list(zip(parameters_per_iter, updated_nodes_per_iter))
+
+
+def extract_parameters(
+    vc_repo: VersionControlForMetaModeling,
+    project_uuid: ProjectID,
+    commit_id: CommitID,
+) -> Parameters:
+    raise NotImplementedError()
+    # TODO: get snapshot
+    # TODO: iter nodes, if iter, read&save outputs
+    # - order?  = order( project_nodes.keys() )
+
+
+# DOMAIN MODEL for project iteration ------------------------------------------------------------
+
+
+class ProjectIteration(BaseModel):
+    """
+    - Keeps a reference of the version: vc repo/commit and workcopy
+    - Keeps a reference of the interation: order, parameters (for the moment, only hash), ...
+
+    - de/serializes from/to a vc tag
+    """
+
+    # version-control info
+    repo_id: Optional[int] = None
+    repo_commit_id: CommitID = Field(...)
+
+    # iteration info
+    iter_index: int = Field(...)
+    total_count: Union[int, Literal["unbound"]] = "unbound"
+    parameters_checksum: SHA1Str = Field(...)
+
+    @classmethod
+    def from_tag_name(
+        cls, tag_name: str, *, return_none_if_fails: bool = False
+    ) -> Optional["ProjectIteration"]:
+        """Parses iteration info from tag name"""
+        try:
+            return cls.parse_obj(parse_iteration_tag_name(tag_name))
+        except ValidationError as err:
+            if return_none_if_fails:
+                log.debug("%s", f"{err=}")
+                return None
+            raise
+
+    def to_tag_name(self) -> str:
+        """Composes unique tag name for this iteration"""
+        return compose_iteration_tag_name(
+            repo_commit_id=self.repo_commit_id,
+            iter_index=self.iter_index,
+            total_count=self.total_count,
+            parameters_checksum=self.parameters_checksum,
+        )
+
+
+def compose_iteration_tag_name(
+    repo_commit_id: CommitID,
+    iter_index: int,
+    total_count: Union[int, str],
+    parameters_checksum: SHA1Str,
+) -> str:
+    """Composes unique tag name for iter_index-th iteration of repo_commit_id out of total_count"""
+    # NOTE: prefers to a json dump because it allows regex search?
+    # should be searchable e.g. get repo/*/*/*
+    # can do the same with json since it is a str!?
+    return (
+        f"iteration:{repo_commit_id}/{iter_index}/{total_count}/{parameters_checksum}"
+    )
+
+
+def parse_iteration_tag_name(name: str) -> Dict[str, Any]:
+    if m := re.match(
+        r"^iteration:(?P<repo_commit_id>\d+)/(?P<iter_index>\d+)/(?P<total_count>-*\d+)/(?P<parameters_checksum>.*)$",
+        name,
+    ):
+        return m.groupdict()
+    return {}
+
+
+# GET/CREATE iterations ------------------------------------------------------------
+
+
+async def get_or_create_runnable_projects(
+    request: web.Request,
+    project_uuid: ProjectID,
+) -> Tuple[List[ProjectID], List[CommitID]]:
```

Comment:
maybe a NamedTuple RunnableProjects ?

Reply:
In this case i think it is more readable this way: it returns a list of project ids and a list of commit ids 
Adding an extra NamedTuple would obscure the readability here because:
 - will have to check the fields 
 - it will exclusively used for this return which btw it is unzipped right away upon call
```python
running_project_ids, project_vc_commits = get_or_create_runnable_projects( ... )
```